### PR TITLE
FEATURE: Allow customization of robots.txt

### PR DESCRIPTION
**Pull request by OsamaSayegh**
This allows admins to customize/override the content of the robots.txt
file at /admin/customize/robots. That page is not linked to anywhere in
the UI -- admins have to manually type the URL to access that page.

Meta topic: https://meta.discourse.org/t/needing-to-edit-robots-txt-file-where-is-it/93879?u=osama

Screenshots:

image

image

@eviltrout does it make sense to prepend a comment to robots.txt that says something along the lines of "this robots.txt file has been customized at /admin/customize/robots" **if** the file is customized? It might help with figuring out why certain things are in the file and how to remove/change them?

### CODE REVIEW

**Comment by coding-horror** [comment]
how does this interact with the handful of site settings that modify `robots.txt`, such as `whitelisted crawler user agents`, `slow down crawler user agents`, and `allow index in robots txt` ?

**Comment by OsamaSayegh** [comment]
@coding-horror this allows overriding the whole robots.txt file. So if an admin goes to this page and makes some changes, then that's what's going to be served. Changes to the site settings that you mentioned won't apply if there is an overriding copy in the database (admins can remove the overriding copy and restore the default robots.txt). Is this how you think it should work?

**Comment by coding-horror** [comment]
It's fine, we just need to make that clear somewhere, also is there a
"restore" button to revert robots.txt to default?

**Comment by OsamaSayegh** [comment]
Yes, there is a "revert changes" button. I will add a copy to the page to make that clear.

File: app/controllers/robots_txt_controller.rb [hunk 1]
```diff
@@ -46,12 +50,13 @@ def index
...
+    result = self.class.fetch_robots_info
+    overridden = SiteSetting.overridden_robots_txt
+    result[:overridden] = overridden if overridden.present?
+    render json: result
   end
 
-protected
-
-  def fetch_robots_info
+  def self.fetch_robots_info
```
Comment (eviltrout) [inline app/controllers/robots_txt_controller.rb]: This method now does not return the robots info, it returns the default robots info (other code is supposed to know about this and return an override if it exists.) I would suggest renaming this to `fetch_default_robots_info`

File: spec/requests/admin/robots_txt_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+  describe "non-admin users" do
+    before { sign_in(user) }
+
+    it "can't see #show" do
+      get "/admin/customize/robots.json"
+      expect(response.status).to eq(404)
+    end
+
+    it "can't see #update" do
```
Comment (eviltrout) [inline spec/requests/admin/robots_txt_controller_spec.rb]:
```suggestion
    it "can't perform #update" do
```

File: spec/requests/admin/robots_txt_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      put "/admin/customize/robots.json", params: { content: "adasdasd" }
+      expect(response.status).to eq(404)
+      expect(SiteSetting.overridden_robots_txt).to eq("")
+    end
+  end
+
+  describe "#show" do
+    before { sign_in(admin) }
+
+    it "returns default content is there are no overrides" do
```
Comment (eviltrout) [inline spec/requests/admin/robots_txt_controller_spec.rb]:
```suggestion
    it "returns default content if there are no overrides" do
```

**Comment by eviltrout** [comment]
> does it make sense to prepend a comment to robots.txt that says something along the lines of "this robots.txt file has been customized at /admin/customize/robots" if the file is customized?

The problem is that information is only really useful to staff. You could put in logic that only includes that message if the request is via a staff account, but some people will use those with API keys to generate the robots.txt for the root of a subfolder site.

File: app/controllers/robots_txt_controller.rb [hunk 4]
```diff
@@ -46,12 +50,13 @@ def index
   # JSON that can be used by a script to create a robots.txt that works well with your
   # existing site.
   def builder
-    render json: fetch_robots_info
+    result = self.class.fetch_robots_info
```
[the remark this answers is not in the document]
Comment (OsamaSayegh) [reply app/controllers/robots_txt_controller.rb]: I'm sorry I'm not sure I understand this point correctly. Do you mean if there is an overridden version, we should try to parse it and send it in JSON format like we do for the default version like this?

```
{
   "header":"# See http://www.robotstxt.org/robotstxt.html ...",
   "agents":[
      {
         "name":"*",
         "disallow":[
            "/auth/",
            "/assets/browser-update*.js",
            ...
         ]
      },
      {
         "name":"mauibot",
         "disallow":[
            "/"
         ]
      },
     ...
   ]
}
```

[the remark this answers is not in the document]
Comment (eviltrout) [reply app/controllers/robots_txt_controller.rb]: Oooh I see the format is much more complicated than I thought. You can disregard my comment, sorry. Let's change the API the way you suggested.

**Pull request merged**